granulate_utils/glogger/__init__.py

```diff
@@ -0,0 +1,189 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+import json
+import logging
+import threading
+import time
+import uuid
+from datetime import datetime
+from logging import Handler, LogRecord
+from typing import List, NamedTuple, Tuple, Union
+
+import backoff
+import requests
+from requests import RequestException
+
+from granulate_utils.glogger.messages_buffer import MessagesBuffer
+
+logger = logging.getLogger(__name__)
+
+
+class Batch(NamedTuple):
+    """A collection of messages sent to the server."""
+
+    logs: List[str]
+    size: int
+    head_serial_no: int
+
+
+class BatchRequestsHandler(Handler):
+    """
+    logging.Handler that accumulates logs in a buffer and flushes them periodically or when
+    threshold is reached. The logs are transformed into dicts. Requests to the server are serialized
+    in a dedicated thread to reduce congestion on the server.
+    """
+
+    scheme = "https"
+
+    # If Tuple[float, float], then the first value is connection-timeout and the second read-timeout.
+    # See https://docs.python-requests.org/en/latest/api/#requests.request
+    request_timeout: Union[float, Tuple[float, float]] = 1.5
+    stop_timeout: float = 60.0
+
+    def __init__(
+        self,
+        server_address: str,
+        max_message_size=1 * 1024 * 1024,  # 1mb
+        max_total_length=5 * 1024 * 1024,  # 5mb
+        flush_interval=10.0,
+        flush_threshold=0.8,
+        overflow_drop_factor=0.25,
+        max_send_tries=5,
+    ):
+        super().__init__(logging.DEBUG)
```

Comment:
Why DEBUG? Allow to control?

Reply:
That's just the default. Can always be controlled using `hdlr.setLevel` if needed. DEBUG is a sane default that we use in gprofiler and elsewhere.